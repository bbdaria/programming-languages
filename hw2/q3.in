print "===TEST START===\n";
val match1 = isMatch "a*b*" "aaabbb";
val match2 = isMatch "ab*c" "abbbc";
val match3 = isMatch "a*b*c" "aabbc";
val match4 = isMatch "abc*" "abc"; 
val match5 = isMatch "abc*" "abccc"; 
val match6 = isMatch "a*b*c*" "aaabbbcccc";
val match7 = isMatch "a*b*c" "aabbcc";
val match8 = isMatch "*abc" "xyzabc";   (* invalid pattern *)
val match9 = isMatch "a*b*" "aabbxyz";
val match10 = isMatch "a*b*c" "c";
val match11 = isMatch "b*" "bbbb";
val match12 = isMatch "b*" "c";
