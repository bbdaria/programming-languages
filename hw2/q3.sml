fun cutLeadingSequence (ch: char) (charList: char list) : char list =
    let
        fun removeLeadingChars [] = []
          | removeLeadingChars (x::xs) =
              if x = ch then removeLeadingChars xs
              else x::xs
    in
        removeLeadingChars charList
    end


type regexp = string

fun isMatch (s1: regexp) (s2: string) : bool =
    let
        val let_arr1 = String.explode s1
        val let_arr2 = String.explode s2

        fun matchLists([], []) = true
          | matchLists(_, []) = false
          | matchLists([], c2::cs2) =
              if not (null cs2) andalso hd cs2 = #"*" then matchLists([], tl cs2) else false
          | matchLists(c1::cs1, c2::cs2) =
              if not (null cs2) andalso hd cs2 = #"*" then
                  matchLists(cutLeadingSequence(c2) (c1::cs1), tl cs2)
              else if c2 = #"*" then
                  matchLists(c1::cs1, cs2)
              else if c1 = c2 then
                  matchLists(cs1, cs2)
              else
                  false
    in
        matchLists(let_arr2, let_arr1)
    end;
