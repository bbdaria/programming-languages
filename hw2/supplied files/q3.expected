val it = () : unit
val match1 = true : bool
val match2 = true : bool
val match3 = true : bool
val match4 = true : bool
val match5 = true : bool
val match6 = true : bool
val match7 = false : bool
val match8 = false : bool
val match9 = false : bool
val match10 = true : bool
val match11 = true : bool
val match12 = false : bool
- 
